ArduCopter/takeoff.cpp

```diff
@@ -201,9 +200,11 @@ void Mode::auto_takeoff_run()
     // call attitude controller with auto yaw
     attitude_control->input_thrust_vector_heading(pos_control->get_thrust_vector(), auto_yaw.get_heading());
 
-    // handle takeoff completion
-    bool reached_altitude = (copter.pos_control->get_pos_target_z_cm() - auto_takeoff_start_alt_cm) >= ((auto_takeoff_complete_alt_cm + terr_offset - auto_takeoff_start_alt_cm) * 0.90);
-    bool reached_climb_rate = copter.pos_control->get_vel_desired_cms().z < copter.pos_control->get_max_speed_up_cms() * 0.1;
+    // takeoff complete when we are less than 1% of the stopping distance from the target altitude
+    // and 10% our maximum climb rate
+    const float vel_threshold_fraction = 0.1;
+    bool reached_altitude = copter.pos_control->get_pos_target_z_cm() >= (auto_takeoff_complete_alt_cm + terr_offset - 0.5 * sq(vel_threshold_fraction * copter.pos_control->get_max_speed_up_cms()) / copter.pos_control->get_max_accel_z_cmss());
```

Comment:
I don't fully understand this change but if auto_takeoff_start_alt_cm is removed I think there must be a change in how auto_takeoff_start() method's "complete_alt_cm" argument is interpreted (aka its frame).  I worry about a built in assumption that the vehicle is taking off when the alt-above-terrain is zero.  This is normally true but not always true (there can be an issue with terrain altitude clamping) meaning that the takeoff does not work (or climbs too high) when this happens.

Reply:
I think we are happy with this now.